src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs

```diff
@@ -407,105 +407,94 @@ internal void NotifyActivityAddException(Activity activity, Exception exception,
         }
     }
 
-    // SynchronizedList<T> is a helper collection which ensure thread safety on the collection
-    // and allow enumerating the collection items and execute some action on the enumerated item and can detect any change in the collection
-    // during the enumeration which force restarting the enumeration again.
-    // Caution: We can have the action executed on the same item more than once which is ok in our scenarios.
+    //this class uses a copy-on-write design to ensure thread safety all operations are thread safe.
+    //However, it is possible for read-only operations to see stale versions of the item while a change
+    //is occurring. This should not be a practical issue if updates are infrequent
     internal sealed class SynchronizedList<T>
     {
-        private readonly List<T> _list;
-        private uint _version;
-
-        public SynchronizedList() => _list = new List<T>();
+        private readonly object _writeLock;
+        //This array must not be mutated directly. To mutate, obtain the lock, copy the array and then replace it with the new array.
+        private T[] _volatileArray;
+        public SynchronizedList()
+        {
+            _volatileArray = [];
+            _writeLock = new();
+        }
 
         public void Add(T item)
         {
-            lock (_list)
+            lock (_writeLock)
```

Comment:
This can be `lock (this)` to save an extra object allocation

Reply:
Sure thing (i thought it was generally frowned upon to `lock(this)`)

Reply:
fixed

Reply:
> Sure thing (i thought it was generally frowned upon to lock(this))

It was always fine on internal types. It was frowned upon in partial trust .NET Framework code.